Source/WTF/wtf/glib/SysprofAnnotator.h

```diff
@@ -0,0 +1,436 @@
+ *  License as published by the Free Software Foundation; either
+ *  version 2 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Library General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Library General Public License
+ *  along with this library; see the file COPYING.LIB.  If not, write to
+ *  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ *  Boston, MA 02110-1301, USA.
+ */
+
+#pragma once
+
+#include <sysprof-capture.h>
+#include <wtf/HashMap.h>
+#include <wtf/Lock.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/Vector.h>
+#include <wtf/text/ASCIILiteral.h>
+
+namespace WTF {
+
+class SysprofAnnotator final {
+    WTF_MAKE_NONCOPYABLE(SysprofAnnotator);
+
+    using RawPointerPair = std::pair<const void*, const void*>;
+    using TimestampAndString = std::pair<int64_t, Vector<char>>;
+
+public:
+
+    static SysprofAnnotator& singleton()
+    {
+        static LazyNeverDestroyed<SysprofAnnotator> instance;
+        return instance;
+    }
+
+    void instantMark(std::span<const char> name, const char* description, ...) WTF_ATTRIBUTE_PRINTF(3, 4)
+    {
+        va_list args;
+        va_start(args, description);
+        sysprof_collector_mark_vprintf(SYSPROF_CAPTURE_CURRENT_TIME, 0, m_processName, name.data(), description, args);
+        va_end(args);
+    }
+
+    void mark(Seconds timeDelta, std::span<const char> name, const char* description, ...) WTF_ATTRIBUTE_PRINTF(4, 5)
+    {
+        va_list args;
+        va_start(args, description);
+        sysprof_collector_mark_vprintf(SYSPROF_CAPTURE_CURRENT_TIME, timeDelta.microsecondsAs<int64_t>(), m_processName, name.data(), description, args);
+        va_end(args);
+    }
+
+    void beginMark(const void* pointer, std::span<const char> name, const char* description, ...) WTF_ATTRIBUTE_PRINTF(4, 5)
+    {
+        auto key = std::make_pair(pointer, static_cast<const void*>(name.data()));
+
+        Vector<char, 1024> buffer;
+        va_list args;
+        va_start(args, description);
+        vsnprintf(buffer.data(), 1024, description, args);
+        va_end(args);
+
+        auto value = std::make_pair(SYSPROF_CAPTURE_CURRENT_TIME, WTFMove(buffer));
+
+        Locker locker { m_lock };
+        m_ongoingMarks.set(key, value);
+    }
+
+    void endMark(const void* pointer, std::span<const char> name, const char* description, ...) WTF_ATTRIBUTE_PRINTF(4, 5)
+    {
+        auto key = std::make_pair(pointer, static_cast<const void*>(name.data()));
+        std::optional<TimestampAndString> value;
+
+        Locker locker { m_lock };
+        if (m_ongoingMarks.contains(key))
+            value = m_ongoingMarks.take(key);
+        locker.unlockEarly();
```

Comment:
```suggestion
        std::optional<TimestampAndString> value;
        {
            Locker locker { m_lock };
            value = m_ongoingMarks.take(key);
        }
```

Reply:
We don't need contains + take, because take already returns the empty value (which for an optional should be nullopt) when the key is not present

Reply:
I'm not sure about removing `m_ongoingMarks.contains(key)`. When I remove it, `m_ongoingMarks.take(key)` returns a non-null `std::pair(0, "")`, which is different from `std::nullptr`. That's because the HashMap stores values of type `std::pair<int64_t, Vector<char>> - not an optional - so the "empty" value is resolved to be `std::pair(0, "")`. In this function, I'm using an `std::optional` on top of the TimestampAndString type.